### DISCUSSION TITLE
How do I simulate a carriage_return? (broken after upgrade)

### DISCUSSION THREAD

**Discussion by beetleb** [xonsh/xonsh#5766] [Q&A] [answered]
In my rc.xsh file I have:

```
try:
    from xonsh.ptk.key_bindings import carriage_return
except ImportError:
    from xonsh.ptk_shell.key_bindings import carriage_return
```

I use `carriage_return` in some of my Xonsh functions. 

I recently upgraded from 0.14.0 to 0.19.0 and this fails to import:

```
No module named 'xonsh.ptk_shell'
```

It seems this was taken out of xonsh and put into its own [package](https://github.com/xonsh/xontrib-ptk-shell) but that repository is archived and the install instructions don't work.

Essentially, I need to simulate pressing a carriage return in a custom keybinding I have. How can I do this?

Thanks.

**Comment by bestlem** [comment]
The link gives a 404 and does not point to the text

**Reply by beetleb** [reply]
Fixed it in the original post, but here it is: https://github.com/xonsh/xontrib-ptk-shell

**Comment by Spitfire1900** [comment]
Import https://github.com/prompt-toolkit/python-prompt-toolkit/blob/master/src/prompt_toolkit/keys.py#L197C5-L197C10 instead, is this to set custom keybindings?

**Comment by anki-code** [comment, marked as answer]
In xonsh 0.19.0 the path to ptk_shell was changed ([src](https://github.com/xonsh/xonsh/blob/5f62cd8b4eaf644d50573809f9f24506eae8df64/xonsh/shells/ptk_shell/key_bindings.py#L32)), try:
```xsh
from xonsh.shells.ptk_shell.key_bindings import carriage_return
```

**Reply by beetleb** [reply]
Thanks!
